napari/utils/events/containers/_nested_list.py

```diff
@@ -0,0 +1,398 @@
+            raise ValueError('The parent index must be a tuple of int')
+        return cast(ParentIndex, tuple(first)), last
+    return ParentIndex(()), -1  # empty tuple appends to self
+
+
+class NestableEventedList(EventedList[_T]):
+    """Nestable Mutable Sequence that emits recursive events when altered.
+
+    ``NestableEventedList`` instances can be indexed with a ``tuple`` of
+    ``int`` (e.g. ``mylist[0, 2, 1]``) to retrieve nested child objects.
+
+    A key property of this class is that when new mutable sequences are added
+    to the list, they are themselves converted to a ``NestableEventedList``,
+    and all of the ``EventEmitter`` objects in the child are connect to the
+    parent object's ``_reemit_nested_event`` method (assuming the child has
+    an attribute called ``events`` that is an instance of ``EmitterGroup``).
+    When ``_reemit_nested_event`` receives an event from a child object, it
+    remits the event, but changes any ``index`` keys in the event to a
+    ``NestedIndex`` (a tuple of ``int``) such that indices emitted by any given
+    ``NestableEventedList`` are always relative to itself.
+
+
+    Parameters
+    ----------
+    data : iterable, optional
+        Elements to initialize the list with. by default None.
+    basetype : type or sequence of types, optional
+        Type of the elements in the list.
+    lookup : dict of Type[L] : function(object) -> L
+        Mapping between a type, and a function that converts items in the list
+        to that type.
+
+    Events
+    ------
+    types used:
+        Index = Union[int, Tuple[int, ...]]
+
+    inserting (index: Index)
+        emitted before an item is inserted at ``index``
+    inserted (index: Index, value: T)
+        emitted after ``value`` is inserted at ``index``
+    removing (index: Index)
+        emitted before an item is removed at ``index``
+    removed (index: Index, value: T)
+        emitted after ``value`` is removed at ``index``
+    moving (index: Index, new_index: Index)
+        emitted before an item is moved from ``index`` to ``new_index``
+    moved (index: Index, new_index: Index, value: T)
+        emitted after ``value`` is moved from ``index`` to ``new_index``
+    changed (index: Index, old_value: T, value: T)
+        emitted when ``index`` is set from ``old_value`` to ``value``
+    changed <OVERLOAD> (index: slice, old_value: List[_T], value: List[_T])
+        emitted when slice ``index`` is set from ``old_value`` to ``value``
+    reordered (value: self)
+        emitted when the list is reordered (eg. moved/reversed).
+    """
+
+    # WAIT!! ... Read the ._list module docs before reimplement these classes
+    # def append(self, item): ...
+    # def clear(self): ...
+    # def pop(self, index=-1): ...
+    # def extend(self, value: Iterable[_T]): ...
+    # def remove(self, value: T): ...
+
+    @overload
+    def __getitem__(self, key: int) -> _T:
+        ...  # pragma: no cover
```

Comment:
Doesn't an int return *either* a `_T` or a `NestableEventedList[_T]`?

This is all super cool btw — I'd seen `overload` but never IRL. Since the implementation is unchanged, I guess this is all only to map input types to output types? Probably worth mentioning in a comment right above. Actually I don't really understand why the implementation falls back to the default one...

Reply:
you're absolutely right.  I think I left it out because mypy yelled at me for it.  here's the error if you're curious:

<img width="693" alt="Screen Shot 2020-08-15 at 6 35 41 PM" src="https://user-images.githubusercontent.com/1609449/90322766-350cbd00-df26-11ea-8142-626891ded894.png">

however, I think in this case, it's probably better to give the proper return type and ignore the mypy error.

> I guess this is all only to map input types to output types?

Yes exactly. `@overload` is amazing and it really works wonders for catching errors.

> Actually I don't really understand why the implementation falls back to the default one...

you mean at runtime?  I think it's just because it's the last one declared right?